### PR TITLE
Add always trigger stop

### LINKED ISSUE #2045: Time-based cover: Allow triggering the stop action even if has_end_stop_sensor is true
**Describe the problem you have/What new integration you would like**
I've a cover with 2 relays: up and down that are managed by a ESP8266.
It has a classic up and down button close to the window for manual operation (for better WAF). 
If I trigger to close the cover with HA, it'll toggle to `CLOSED` state (as expected). However, when someone use the physical button to open it, HA isn't aware of it and nor is ESPHome. 
So the next time I use the "close" action in HA, nothing happens and the cover doesn't close (unless I use `has_end_stop_sensor` variable).
If I'm using the `has_end_stop_sensor` the cover will close (as expected), but the relay remained powered after the clover is closed because the stop action isn't called. So if someone press the physical "open" button, it'll have both phases on the motor which isn't good (on my motor, it triggers the programming mode where it sets the new end/stop position, but on many system it's forbidden, it'll trip the fuse).

So ideally, I'd like to be able to run the stop action (that's releasing the relay) after any action is performed (open or closed) **and** use the `has_end_stop_sensor` option since HA can't really know the position of the cover and only get an estimate of it.

**Please describe your use case for this integration and alternatives you've tried:**
If I don't use the `cover` type but a plain button, I can set a time based button to open or close the cover, but then HA doesn't know it's a cover, and there's no tracking of the cover state. So in that case, I can't set the cover midway. If I don't add a delay, then I have to use a switch and it requires manipulating the interface twice without forgetting to release the relay.

**Additional context**
The code that would need to change is in `time_based_cover.cpp`, around line 34, that should read:
```cpp
  if (this->is_at_target_()) {
    if (this->has_built_in_endstop_ && !this->always_trigger_stop_ &&
        (this->target_position_ == COVER_OPEN || this->target_position_ == COVER_CLOSED)) {
      // Don't trigger stop, let the cover stop by itself.
      this->current_operation = COVER_OPERATION_IDLE;
    } else {
      this->start_direction_(COVER_OPERATION_IDLE);
    }
```
And a new configuration option `always_trigger_stop` should be added to the class.

### PR DESCRIPTION
**Pull request by X-Ryl669**
# What does this implement/fix?

This should fix [#2045](https://github.com/esphome/feature-requests/issues/2045)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes esphome/feature-requests#2045

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2571

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [X] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
- platform: time_based
  name: "Kitchen"
  id: cover_1
  open_action:
    - switch.turn_on: relay_kitchen_up
  open_duration: 20s
  close_action:
    - switch.turn_on: relay_kitchen_down
  close_duration: 20s
  stop_action:
    - switch.turn_off: relay_kitchen_up
    - switch.turn_off: relay_kitchen_down
  manual_control: true
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by X-Ryl669** [comment]
Ready for merge.

**Comment by nagyrobi** [comment]
In what aspect does this relate to https://github.com/esphome/esphome/pull/2045 ?

**Comment by X-Ryl669** [comment]
Not this one. [This one](https://github.com/esphome/feature-requests/issues/2045).
If I understand correctly, feature requests are in their own repository now.

**Comment by X-Ryl669** [comment]
Fixed the link too here.

File: esphome/components/time_based/time_based_cover.cpp [hunk 1]
```diff
@@ -79,6 +79,14 @@ void TimeBasedCover::control(const CoverCall &call) {
     auto pos = *call.get_position();
     if (pos == this->position) {
       // already at target
+      if (this->manual_control_ && (pos == COVER_OPEN || pos == COVER_CLOSED)) {
+        // for covers with manual control switch, we can't rely on the computed position, so if
+        // the command triggered again, we'll assume it's in the opposite direction anyway.
+        auto op = pos == COVER_CLOSED ? COVER_OPERATION_CLOSING : COVER_OPERATION_OPENING;
+        this->position = pos == COVER_CLOSED ? 1.0f : 0.0f;
```
Comment (glmnet) [inline esphome/components/time_based/time_based_cover.cpp]: can this line better be:
```suggestion
        this->position = pos == COVER_CLOSED ? COVER_OPEN : COVER_CLOSED;
```

Comment (X-Ryl669) [reply esphome/components/time_based/time_based_cover.cpp]: Right! Thanks for spotting it.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**